### PR TITLE
24HFwXaC Assume roles locally (frontend)

### PR DESCRIPTION
**Pull request by j-maynard**
### What
A page (profile) where the dev users can assume different roles as defined in the previous card. We can assume multiple roles (check boxes).

### Why
This will enable us testing different roles.

### Whatelse
This PR also has some small changes to pre-commit to help it run better and some tweaks to the previous backend work.

### CODE REVIEW

File: app/policies/profile_controller_policy.rb [hunk 1]
```diff
@@ -2,4 +2,16 @@ class ProfileControllerPolicy < ApplicationPolicy
   def show?
     true
   end
+
+  def switch_client?
```
Comment (jkunle) [inline app/policies/profile_controller_policy.rb]:
```
  def switch_client?
    !Rails.env.development?
  end

  def update_role?
    !Rails.env.development?
  end
```

File: config/locales/en.yml [hunk 2]
```diff
@@ -89,6 +89,11 @@ en:
     permissions: "Permissions"
     permissions_heading: "Your Permissions"
     access: "Access"
+    client_switch_h: "Switch Authentication Clients"
+    client_switch_l: "In development you can switch between using cognito and the stub authenication clients. If you use the stub client you can assume different roles."
```
Comment (jkunle) [inline config/locales/en.yml]: authentication

File: app/controllers/profile_controller.rb [hunk 3]
```diff
@@ -1,5 +1,36 @@
...
+      flash[:notice] = 'Switched to Stub Auth Client'
+    end
+    redirect_to profile_path
+  end
+
+  def update_role
+    role_str = params[:assume_roles][:role_selection].join(',')
+    if params[:assume_roles][:role_selection].include?(ROLE::GDS)
+      CognitoStubClient.update_user(
+        role: role_str, email_domain: 'digital.cabinet-office.gov.uk'
```
Comment (jkunle) [inline app/controllers/profile_controller.rb]: Do we need a constant for digital.cabinet-office.gov.uk ?

File: config/routes.rb [hunk 4]
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
```
Comment (brenetic) [inline config/routes.rb]: ~# frozen_string_literal: true~

File: app/controllers/profile_controller.rb [hunk 5]
```diff
@@ -1,5 +1,36 @@
 class ProfileController < ApplicationController
   def show
+    if Rails.env.development?
+      @stub_available = true
+      @using_stub = CognitoStubClient.true?(SelfService.service(:cognito_stub))
+      @breakerofchains = @using_stub && current_user.given_name == 'Daenerys'
```
Comment (brenetic) [inline app/controllers/profile_controller.rb]: I think this isn't Easter Egg hidden enough . . . .

File: app/views/profile/show.html.erb [hunk 6]
```diff
@@ -1,3 +1,18 @@
+<% if !Rails.env.production? %>
```
Comment (Rosa-Fox) [inline app/views/profile/show.html.erb]: `unless Rails.env.production?`

File: app/controllers/profile_controller.rb [hunk 7]
```diff
@@ -1,5 +1,36 @@
...
+      CognitoStubClient.update_user(
+        role: role_str, email_domain: 'digital.cabinet-office.gov.uk'
+      )
+    else
+      CognitoStubClient.update_user(role: role_str)
+    end
+    UserSignOutEvent.create(user_id: warden.user.user_id)
+    signed_out = (Devise.sign_out_all_scopes ? sign_out : sign_out(resource_name))
+    flash[:notice] = 'You need to sign in again for role changes to take effect' if signed_out
+    redirect_to root_path
```
Comment (brenetic) [inline app/controllers/profile_controller.rb]: The flash messages need to be locales

File: config/routes.rb [hunk 8]
```diff
@@ -1,3 +1,5 @@
...
+
```
Comment (brenetic) [inline config/routes.rb]: ~# frozen_string_literal: true~

File: config/locales/en.yml [hunk 9]
```diff
@@ -89,6 +89,11 @@ en:
     permissions: "Permissions"
     permissions_heading: "Your Permissions"
     access: "Access"
+    client_switch_h: "Switch Authentication Clients"
```
Comment (Rosa-Fox) [inline config/locales/en.yml]: maybe the `h` and `l` could be more descriptive?

File: lib/auth/cognito_stub_client.rb [hunk 10]
```diff
@@ -41,21 +46,47 @@ def self.stub_client
...
+    return "in prod" if Rails.env.production?
+    # Exit if we're already a stub client
+    return "already stub" unless false?(SelfService.service(:cognito_stub))
+
+    real_client = SelfService.service(:cognito_client)
+    SelfService.register_service(name: :real_client, client: real_client)
+    SelfService.register_service(name: :cognito_client, client: stub_client)
+    setup_stubs
+    SelfService.register_service(name: :cognito_stub, client: 'true')
+    "switched to stub"
```
Comment (brenetic) [inline lib/auth/cognito_stub_client.rb]: These return strings aren't used for `switch_to_stub` and `switch_to_cognito`. The flash messages are set independently above

File: lib/auth/cognito_stub_client.rb [hunk 11]
```diff
@@ -41,21 +46,47 @@ def self.stub_client
...
     end
   end
+
+  def self.true?(obj)
+    obj.to_s.downcase == 'true'
+  end
+
+  def self.false?(obj)
+    obj.to_s.downcase == 'false'
+  end
```
Comment (brenetic) [inline lib/auth/cognito_stub_client.rb]:
```suggestion
def self.using_coginito_stub?(obj)
  obj.to_s.downcase == 'true'
end
```

`!self.using_cognito_stub(SelfService.service(:cognito_stub)` would be fine

File: config/locales/en.yml [hunk 12]
```diff
@@ -89,6 +89,11 @@ en:
...
+    client_switch_l: "In development you can switch between using cognito and the stub authentication clients. If you use the stub client you can assume different roles."
+    banner: "There are features on this page which are only avialable in the development environment."
+    assume_roles_h: "Switch Roles"
+    assume_roles_l: "Select which role(s) you wish to assume?"
```
Comment (brenetic) [inline config/locales/en.yml]: What is `l` and `h` referring to?

File: spec/controllers/profile_controller_spec.rb [hunk 13]
```diff
@@ -16,5 +16,45 @@
...
+      @user = FactoryBot.create(:user)
+      allow(request.env['warden']).to receive(:authenticate!).and_return(@user)
+      allow(controller).to receive(:current_user).and_return(@user)
+      sign_in @user
+      get :show
+      expect(@controller.instance_variable_get(:@stub_available)).to eq(true)
+      expect(@controller.instance_variable_get(:@breakerofchains)).to eq(false)
+      expect(@controller.instance_variable_get(:@using_stub)).to eq(true)
+      Rails.env = 'test'
+    end
```
Comment (brenetic) [inline spec/controllers/profile_controller_spec.rb]: These can be  wrapped in their own context.

Possibly two separate contexts. If so, then `after_each` can be used to reset the Rails.env to test

**Pull request merged**